### ISSUE TITLE
summary_image_op_test failure on ppc64le

### ISSUE THREAD

**Issue by vaibhavsood** [tensorflow/tensorflow#12325]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: SLES/RHEL ppc64le
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: master
- **Python version**: 2.7.4
- **Bazel version (if compiling from source)**: 0.4.5
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

bazel test **-c opt** --local_resources 4096,4.0,1.0 -j 1 --test_output=errors //tensorflow/python/kernel_tests:summary_image_op_test

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

This test fails when run with -c opt, it passes with dbg & fastbuild. Recently there was a commit (link below) adjusting the tolerance value to pass this test, it however fails for 'opt' build even after the change. 

Issue: https://github.com/tensorflow/tensorflow/issues/11967
Commit fixing this:  https://github.com/tensorflow/tensorflow/commit/901a3b5014e2061e16a251cda24de44179d7c21a

Is it okay to ignore the 'opt' build failure or should this issue be root-caused further? thanks!

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

...
Not equal to tolerance rtol=2e-05, atol=2e-05
None
(mismatch 2.85714285714%)
 x: array([[[ 221.],
        [ 102.],
        [ 129.],...
 y: array([[[221],
        [102],
        [129],...

----------------------------------------------------------------------
Ran 3 tests in 0.052s

FAILED (failures=1)
not close where =  (array([3]), array([6]), array([0]))
not close lhs =  [0]
not close rhs =  [ 1.]
not close dif =  [ 1.]
not close tol =  [  3.99999990e-05]
dtype = uint8, shape = (5, 7, 1)
================================================================================
Target //tensorflow/python/kernel_tests:summary_image_op_test up-to-date:
  bazel-bin/tensorflow/python/kernel_tests/summary_image_op_test
INFO: Elapsed time: 9.223s, Critical Path: 1.12s
//tensorflow/python/kernel_tests:summary_image_op_test                   FAILED
  /root/.cache/bazel/_bazel_root/55751bac5fd009d8cf4f164f81814728/execroot/tenso

Executed 1 out of 1 test: 1 fails locally.

**Comment by skye** [comment]
I don't think the TF team officially supports ppc64le. @martinwicke, @gunan correct me if I'm wrong. I'll mark this community support.

**Comment by gunan** [comment]
Correct, ppc64le support is best effort.
@sandipmgiri FYI

**Comment by sandipmgiri** [comment]
The test is failing due to a minor array mismatch error (the value of only 1 out of 34 entries differing i.e. 0 vs expected 1) at line number 83 in file https://github.com/tensorflow/tensorflow/blob/v1.3.1/tensorflow/python/kernel_tests/summary_image_op_test.py#L83

I have checked the output of all the functions which are involved in this test, and found that -
Following line returning incorrect result on ppc64le- https://github.com/tensorflow/tensorflow/blob/v1.3.1/tensorflow/python/kernel_tests/summary_image_op_test.py#L73
_**summ = summary.image("img", const)**
value = sess.run(summ)
print ("value")_

Output (X86 vs ppc64le) :
X86 :
```
'\n\x81\x01\n\x0bimg/image/0"r\x08\x05\x10\x07\x18\x01"j\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x001IDAT\x08\x99c\xbc\xdb)-\xbaO\xe7?\xa3\xdc,\xf78I\xb5\xab\xcc\xc9\xb3\xd89\x05\xa7\xf52\x89Y\x07}_`\xa7\xc6d\xf4/\xcd\xee\xab\t\x0f\x00\x1a\xe4\r\xe2\xc3\xbd-\xa5\x00\x00\x00\x00IEND\xaeB`\x82\n\x81\x01\n\x0bimg/image/1"r\x08\x05\x10\x07\x18\x01"j\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x001IDAT\x08\x99c\x8e\xe2\xdbZ\xf5\xa7R\x8e\xc9M\xca\xd7\\#@\x8aYh"\x87\x11\xe3\xa6\xd7\xcc\x7f\xf9\x9f\xb9\x1c^l\xc9\xf8\xa0\xeb\xf1{\x0e\xeds\x00\x0c\xef\x0f:\xdfIp\x07\x00\x00\x00\x00IEND\xaeB`\x82\n\x83\x01\n\x0bimg/image/2"t\x08\x05\x10\x07\x18\x01"l\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x003IDAT\x08\x99\x01(\x00\xd7\xff\x01\xd4\x9c\x16J\x04\xa0\x1e\x04\xd5\x05\x96\x1e\x0e#\xb5\x03 \xea2?\x18\xef&\x03l\xfd\x04\xb9\xe5o\xe2\x02\xc6\xe6\xd4;\x1a/\xeb\x1f3\x10\x07\xa8\xd8;Q\x00\x00\x00\x00IEND\xaeB`\x82'
```
ppc64le:
```
'\n\x81\x01\n\x0bimg/image/0"r\x08\x05\x10\x07\x18\x01"j\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x001IDAT\x08\x99c\xbc\xdb)-\xbaO\xe7?\xa3\xdc,\xf78I\xb5\xab\xcc\xc9\xb3\xd89\x05\xa7\xf52\x89Y\x07}_`\xa7\xcad\xf4/\xcd\xee\xab\t/\x00\x1a\xdc\r\xe2\xc4\x9c\r\x96\x00\x00\x00\x00IEND\xaeB`\x82\n\x81\x01\n\x0bimg/image/1"r\x08\x05\x10\x07\x18\x01"j\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x001IDAT\x08\x99c\x8e\xe2\xdbZ\xf5\xa7R\x8e\xc9M\xca\xd7\\#@\x8aYh"\x87\x11\xe3\xa6\xd7\xcc\x7f\xf9\x9f\xb9\x1c^l\xc9\xf8\xa0\xeb\xf1{\x0e\xeds\x00\x0c\xef\x0f:\xdfIp\x07\x00\x00\x00\x00IEND\xaeB`\x82\n\x83\x01\n\x0bimg/image/2"t\x08\x05\x10\x07\x18\x01"l\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x07\x00\x00\x00\x05\x08\x00\x00\x00\x00\xac\xf1\xa9\x04\x00\x00\x003IDAT\x08\x99\x01(\x00\xd7\xff\x01\xd4\x9c\x16J\x04\xa0\x1e\x04\xd5\x05\x96\x1e\x0e#\xb5\x03 \xea2?\x18\xef&\x03l\xfd\x04\xb9\xe5o\xe2\x02\xc6\xe6\xd4;\x1a/\xeb\x1f3\x10\x07\xa8\xd8;Q\x00\x00\x00\x00IEND\xaeB`\x82'
```
Compared the x86 and ppc64le outputs, difference -
```
**X86**                          vs               **ppc64le**
xc6d                                                     xcad
\t\x0f\                                                   \t/\
a\xe4                                                    a\xdc
xc3\xbd-\xa5                                       xc4\x9c\r\x96
```
For testing purpose , I have replaced these 4 ppc64le values with x86 and the test passed successfully (on power).
What would be the reason of getting `summary.image("img", const) ` output incorrect/different on power?
Please provide any suggestions/comments on this, that will be helpful for further debugging. Thanks!!!
```
Where i/p,
const  =  array([[[[  1.69052565e+00],
         [ -4.65937376e-01],
         [  3.28201652e-02],
         [  4.07516271e-01],
         [ -7.88923025e-01],
         [  2.06557289e-03],
         [ -8.90385883e-04]],

        [[ -1.75472426e+00],
         [  1.01765800e+00],
         [             nan],
         [ -6.25428975e-01],
         [ -1.71548262e-01],
         [  5.05299389e-01],
         [ -2.61356413e-01]],

        [[ -2.42749080e-01],
         [ -1.45324147e+00],
         [  5.54580331e-01],
         [  1.23880908e-01],
         [  2.74459928e-01],
         [ -1.52652454e+00],
         [  1.65069973e+00]],

        [[  1.54335529e-01],
         [ -3.87139946e-01],
         [  2.02907228e+00],
         [ -4.53860313e-02],
         [ -1.45067871e+00],
         [ -4.05227870e-01],
         [ -2.28831506e+00]],

        [[  1.04939651e+00],
         [ -4.16474313e-01],
         [ -7.42553532e-01],
         [  1.07247019e+00],
         [ -1.65107560e+00],
         [  5.35429358e-01],
         [ -2.06441474e+00]]],


       [[[ -6.62159324e-01],
         [ -1.20421982e+00],
         [  1.46197557e+00],
         [  1.76616085e+00],
         [ -3.29413742e-01],
         [  8.40733230e-01],
         [ -1.79986402e-01]],

        [[  5.68061888e-01],
         [ -7.52837181e-01],
         [ -1.70833921e+00],
         [ -1.80309868e+00],
         [  3.83121848e-01],
         [  2.24759507e+00],
         [  2.69411623e-01]],

        [[ -5.24604619e-01],
         [  1.91201890e+00],
         [  2.37301841e-01],
         [  1.01433985e-01],
         [  2.52577722e-01],
         [ -1.32377192e-01],
         [ -3.09476346e-01]],

        [[ -1.43496346e+00],
         [  5.01624107e-01],
         [ -9.47754532e-02],
         [  1.19308591e+00],
         [ -3.68818462e-01],
         [ -1.90636992e+00],
         [ -9.96106341e-02]],

        [[  1.69953728e+00],
         [ -3.83423120e-01],
         [ -8.89856875e-01],
         [ -1.19359195e+00],
         [ -1.05001676e+00],
         [ -3.00193727e-01],
         [ -1.17998207e+00]]],


       [[[  1.49763906e+00],
         [ -2.82635242e-01],
         [  1.08648375e-01],
         [  1.43823957e+00],
         [  1.50331855e+00],
         [ -2.12732971e-01],
         [  3.31974208e-01]],

        [[  7.35026598e-01],
         [ -1.92855462e-01],
         [ -1.77801287e+00],
         [  6.54705703e-01],
         [  8.94352317e-01],
         [  4.15502608e-01],
         [ -9.23544645e-01]],

        [[ -1.96027309e-01],
         [ -5.90769827e-01],
         [ -2.99711227e-01],
         [  1.29688525e+00],
         [  1.52957964e+00],
         [  6.69418216e-01],
         [  5.48745096e-01]],

        [[  6.76629007e-01],
         [ -1.22421868e-02],
         [ -7.56634623e-02],
         [ -6.73645198e-01],
         [ -5.58674484e-02],
         [  2.25994706e+00],
         [  8.69039357e-01]],

        [[ -3.42117012e-01],
         [ -4.71926659e-01],
         [ -8.64489973e-01],
         [  3.74370337e-01],
         [  3.91545683e-01],
         [ -1.44312167e+00],
         [  4.86335188e-01]]],


       [[[ -5.69472253e-01],
         [  1.42672122e+00],
         [  1.56843990e-01],
         [  1.71773005e+00],
         [ -4.58126783e-01],
         [ -2.87983865e-01],
         [  2.99808353e-01]],

        [[  1.05594826e+00],
         [  5.65882504e-01],
         [ -1.23352396e+00],
         [  1.82900697e-01],
         [  2.22447384e-02],
         [ -4.29068595e-01],
         [ -6.48105264e-01]],

        [[  1.74757719e+00],
         [ -3.90386492e-01],
         [ -8.45922649e-01],
         [  6.37112856e-01],
         [  1.30622983e-01],
         [ -7.58138672e-02],
         [  7.81301796e-01]],

        [[  4.88625407e-01],
         [  3.62190008e-01],
         [  9.64200497e-01],
         [  2.83636421e-01],
         [ -6.16921961e-01],
         [ -3.62282217e-01],
         [ -5.20972311e-01]],

        [[  2.09722057e-01],
         [ -1.07798493e+00],
         [ -1.83242595e+00],
         [  8.81549269e-02],
         [ -1.33652055e+00],
         [ -1.95197809e+00],
         [  3.73330772e-01]]]])
```

**Comment by gunan** [comment]
Looks like the images generated are actually random. So I would expect the summaries generated to be slightly different: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/summary_image_op_test.py#L60
However, the random functions are seeded in two places: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/summary_image_op_test.py#L53 https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L618
Maybe the test is actually flaky and the random functions, even when seeded with the same numbers, generate different sequences on x86 and ppc64? Or ordering of ops are different between ppc64 and x86 so we end up with different random images.

So next I tried removing the random seedsand running the test multiple times on x86 to see how flaky it is. I tried the command:
```
$ bazel test -c opt tensorflow/python/kernel_tests:summary_image_op_test --runs_per_test=1000
INFO: Analysed target //tensorflow/python/kernel_tests:summary_image_op_test (0 packages loaded).
INFO: Found 1 test target...
Target //tensorflow/python/kernel_tests:summary_image_op_test up-to-date:
  bazel-bin/tensorflow/python/kernel_tests/summary_image_op_test
INFO: Elapsed time: 38.523s, Critical Path: 1.86s
INFO: Build completed successfully, 1001 total actions
//tensorflow/python/kernel_tests:summary_image_op_test                   PASSED in 1.8s
  Stats over 1000 runs: max = 1.8s, min = 0.7s, avg = 1.2s, dev = 0.1s

Executed 1 out of 1 test: 1 test passes.
```

Can someone try the above on a ppc64le machine?

**Comment by sandipmgiri** [comment]
Hi @gunan,
As you suggested , I tried removing the random seeds and running the test multiple times on ppc64le : 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/summary_image_op_test.py#L53
`# np.random.seed(7)`
```
$ bazel test -c opt tensorflow/python/kernel_tests:summary_image_op_test --runs_per_test=1000
INFO: Found 1 test target...
Target //tensorflow/python/kernel_tests:summary_image_op_test up-to-date:
  bazel-bin/tensorflow/python/kernel_tests/summary_image_op_test
INFO: Elapsed time: 120.027s, Critical Path: 1.15s
//tensorflow/python/kernel_tests:summary_image_op_test                   PASSED in 1.1s
  Stats over 1000 runs: max = 1.1s, min = 0.9s, avg = 0.9s, dev = 0.0s

Executed 1 out of 1 test: 1 test passes.
```

Thanks!

**Comment by gunan** [comment]
This is really interesting :)
OK, so let's remove line 53 from summary image op test if that fixes the test on ppc64le. As far as I am concerned we already seed all random calls in testcasse class.

**Comment by sandipmgiri** [comment]
Sure @gunan , I will create a PR to upstream the changes. Thanks a lot for the suggestions/comments!

**Comment by sandipmgiri** [comment]
Fixed in PR #15147
